DataDog/dd-trace-rb
# Expand aarch64 support to all CI images

**What does this PR do?**

Add first-class support for development on `aarch64-linux`.

**Motivation**

Working from `arm64-darwin` machines becomes an increasingly common thing, and presumably we want to be able to run all Ruby versions on that.

Our `docker-compose` relies on images that aren't currently built for `aarch64-linux`, this commit changes that.

**Additional Notes**

Special cases:

* Removed `protoc` from all `Dockerfile`s since it appears to be unused now
* `ruby:2.1` has no aarch64 image published, Rebuilt it from a Debian Stretch base
* `ruby:2.2` has an aarch64 image published but is Debian Jessie based, and Jessie had aarch64 upon release but as an "extra" arch, which is not part of LTS, so packages vanished when Debian moved the repos to `archive`, so any `apt-get` irrevocably fails. Rebuilt from a Debian Stretch base
* JRuby 9.2 has no aarch64 image published, was based on `openjdk:8-jre`, I rebuilt both from the last `openjdk:8-jre` which now has aarch64 images, these being based on bullseye (there are also Buster-based images upstream if it turns out it matters)
* TruffleRuby 21 has no aarch64 image published, and it wasn't an official one anyway. Added TruffleRuby 22 from the official Oracle registry, and removed TruffleRuby 21

The last bit of adjustment is Stretch-based ones (so 2.1, 2.2, 2.3, 2.4) which do build until they fail to run the first aarch64 release of `docker-compose` due to an outdated glibc version. This has been worked around by installing it from `stretch-backports`, which has a 1.21 version close to 1.29 we presumably want.

2.4 had a note about staying on Stretch instead of Buster because of `sequel`, which may not be relevant anymore. It was initially moved to Buster to work around the `docker-compose` failure, but is currently moved back to Stretch using the above backport alternative.

In all cases attempts were made to stay as close as possible to the original `Dockerfile`s for each version and their base. When things are made to differ it's because there seems to be no realistic alternative. There are opportunistic moves to streamline and clean all of this but this was kept out of scope of this commit to maximise consistency and ease of reviewing.

All images are now building and running for both `x86_64` and `aarch64`.

**How to test the change?**

- Build locally, see `.circleci/images/primary/README.md`
- On an arm64 machine (or an x86_64 machine using qemu via `https://github.com/tonistiigi/binfmt`), and specifying the platform to docker with `--platform linux/aarch64`
- Final testing needs publishing images for CircleCI to pick them up, so results there won't be relevant. But see below.

A subsequent commit will also add a GitHub Workflow that will build these images in CI, with an option to publish them on the GitHub Container Registry at `ghcr.io` for everyone (including CircleCI) to use.
